File: packages/typescript/src/services/rag.ts
import { BaseService } from './base.js';

export interface RAGDocument {
  doc_id: string;
  filename: string;
  uploaded_at: string;
  size: number;
  status: 'indexed' | 'pending' | 'failed';
  chunk_count: number;
}

export interface RAGQueryResult {
  results: Array<{
    doc_id: string;
    chunk_id: string;
    content: string;
    similarity_score: number;
    metadata?: Record<string, unknown>;
  }>;
  total_results: number;
}

export interface RAGUploadResult {
  doc_id: string;
  filename: string;
  size: number;
  status: string;
  chunks: number;
}

export class RAGService extends BaseService {
  async upload(file: Blob, options?: { filename?: string }): Promise<RAGUploadResult> {
    const formData = new FormData();
    formData.append('file', file, options?.filename);
    return this.requestMultipart<RAGUploadResult>('/v1/rag/upload', formData);
  }

  async query(
    query: string,
    options?: { doc_ids?: string[]; top_k?: number },
  ): Promise<RAGQueryResult> {
    return this.request<RAGQueryResult>('POST', '/v1/rag/query', {
      body: { query, ...options },
    });
  }

  async documents(): Promise<{ documents: RAGDocument[]; total: number }> {
    return this.request('GET', '/v1/rag/documents');
  }

  async deleteDocument(docId: string): Promise<{ success: boolean }> {
    return this.request('DELETE', `/v1/rag/documents/${docId}`);
  }
}

File: packages/python/scalix/services/rag.py
"""RAG service — document upload, query, and management."""

from __future__ import annotations

from typing import Any, BinaryIO

from scalix.services.base import BaseService


class RAGService(BaseService):
    async def upload(
        self, file: BinaryIO, *, filename: str | None = None
    ) -> dict[str, Any]:
        files = {"file": (filename, file) if filename else file}
        return await self._request_multipart("/v1/rag/upload", files=files)

    async def query(
        self,
        query: str,
        *,
        doc_ids: list[str] | None = None,
        top_k: int | None = None,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {"query": query}
        if doc_ids:
            body["doc_ids"] = doc_ids
        if top_k:
            body["top_k"] = top_k
        return await self._request("POST", "/v1/rag/query", json=body)

    async def documents(self) -> dict[str, Any]:
        return await self._request("GET", "/v1/rag/documents")

    async def delete_document(self, doc_id: str) -> dict[str, Any]:
        return await self._request("DELETE", f"/v1/rag/documents/{doc_id}")
